include/miopen/miopen.h

```diff
@@ -4666,6 +4666,120 @@ miopenReduceTensor(miopenHandle_t handle,
 /** @} */
 // CLOSEOUT TensorReduce DOXYGEN GROUP
 
+/**
+ * An miopenProblem will describe a problem for different miopen operations.
+ * For now, this is only used for convolution, but could be used for other
+ * operators in the future(such as GEMM, Pooling, etc)
+ */
+MIOPEN_DECLARE_OBJECT(miopenProblem);
+
+typedef enum
+{
+    miopenProblemDirectionForward        = 0,
+    miopenProblemDirectionBackward       = 1,
+    miopenProblemDirectionBackwardWeight = 2,
+} miopenProblemDirection_t;
+
+// Enums with no meaningful default value should probably have invalid default for fool-proofing.
+typedef enum
+{
+    miopenTensorNameInvalid  = 0,
+    miopenTensorConvolutionX = 1,
+    miopenTensorConvolutionW = 2,
+    miopenTensorConvolutionY = 3,
+} miopenTensorName_t;
+
+typedef enum
+{
+    miopenSearchResultsOrderByTime   = 0,
+    miopenSearchResultsOrderByMemory = 1,
+} miopenSearchResultsOrder_t;
+
+typedef enum
+{
+    miopenSearchOptionInvalid          = 0,
+    miopenSearchOptionExhaustiveSearch = 1, // int
+    miopenSearchOptionResultsOrder     = 2, // miopenSearchResultsOrder_t
+    miopenSearchOptionWorkspaceLimit   = 3, // size_t
+} miopenSearchOptionName_t;
+
+typedef enum
+{
+    miopenSolutionAttributeInvalid       = 0,
+    miopenSolutionAttributeWorkspaceSize = 1, // size_t
+    miopenSolutionAttributeTime          = 2, // float
+} miopenSolutionAttribute_t;
+
+miopenStatus_t miopenCreateProblem(miopenProblem_t* problem);
+miopenStatus_t miopenDestroyProblem(miopenProblem_t problem);
+
+// Universal problem should have named tensor fields rather than ordered.
+// Previously, different directions had different input orders, which may lead to confusion.
+// Implicitly ordered fields would also lead to a whole lot of magical ids all over the internals.
+miopenStatus_t miopenSetProblemTensorDescriptor(miopenProblem_t problem,
+                                                miopenTensorName_t name,
+                                                const miopenTensorDescriptor_t descriptor);
+
+miopenStatus_t miopenSetProblemOperatorDescriptor(miopenProblem_t problem,
+                                                  const void* operatorDesc,
+                                                  miopenProblemDirection_t direction);
+
+/**
+ * The miopenSearchOptions will allow the user to configure how search will be
+ * used. This can include things like exhaustiveSearch, and ordering. In the
+ * future, this can support setting different find mode options.
+ */
+MIOPEN_DECLARE_OBJECT(miopenSearchOptions);
+
+miopenStatus_t miopenCreateSearchOptions(miopenSearchOptions_t* options);
+miopenStatus_t miopenDestroySearchOptions(miopenSearchOptions_t options);
+miopenStatus_t miopenSetSearchOption(miopenSearchOptions_t options,
+                                     miopenSearchOptionName_t optionName,
+                                     size_t valueSize,
+                                     void* value);
```

Comment:
You should use different functions for each attribute as using `void*` is error-prone.

Reply:
We could also have search options in the future that takes multiple values.